### PR TITLE
Add CAD (DWG) driver

### PR DESCRIPTION
**Pull request by BishopGIS**
Merge work from GSoC 2016. 
For test if GDAL CI passing.

### CODE REVIEW

File: gdal/configure.in [hunk 1]
```diff
@@ -111,13 +111,23 @@ AX_CHECK_COMPILE_FLAG([-fno-operator-names], [CXX_WFLAGS="$CXX_WFLAGS -fno-opera
...
+fi
+
 dnl Enable -Wimplicit-fallthrough (clang specific) only if C++11 is enabled since CPL_FALLTHROUGH is only active then
 AC_LANG_PUSH([C++])
 SAVED_CXXFLAGS=$CXXFLAGS
 CXXFLAGS="$CXXFLAGS $ERROR_ON_UNKNOWN_OPTIONS -Wimplicit-fallthrough"
 AC_MSG_CHECKING([if -Wimplicit-fallthrough can be enabled])
 AC_COMPILE_IFELSE([AC_LANG_PROGRAM(
-  [[#if __cplusplus < 201103L || !(defined(__clang__) && (__clang_major__ > 3 || (__clang_major__ == 3 && __clang_minor__ >=7)))
+  [[#if __cplusplus < 201103L || (defined(__clang__) && (__clang_major__ < 3 || (__clang_major__ == 3 && __clang_minor__ < 7)))
```
Comment (rouault) [inline gdal/configure.in]: Why was this changed ? The new version will succeed with G++ in C++11 mode whereas it should fail since  -Wimplicit-fallthrough is clang specific

**Comment by rouault** [comment]
You could also try forking https://github.com/rouault/gdal_coverage/tree/trunk_ubuntu16_04 and merging the branch in it so that it is actually compiled with a C++11 compiler to see if it raises some warning. And this one as well https://github.com/rouault/gdal_coverage/tree/trunk_gcc5.2_sanitize.
(I guess we should rather have a test matrix in the main .travis.yml instead of this cumbersome process).

File: gdal/configure.in (hunk 1 above)
Comment (BishopGIS) [inline gdal/configure.in]: I tested on g++ and clang this code. Both versions are support c++11, but failed as this #if is wrong (#if __cplusplus < 201103L || !(defined(__clang__) && (__clang_major__ > 3 || (__clang_major__ == 3 && __clang_minor__ >=7)))) or #error message is wrong. 
Anyhow on g++ configure failed as -Wimplicit-fallthrough is clang specific - is this expected behaviour? 

This is not connected with CAD driver - so I can revert this change.

**Comment by BishopGIS** [comment]
I test this PR on my ubuntu 16.10. Also clang from 3.3 (this travis-ci report clang version 3.4) also support c++11. There is errors in my configure.in which I'm fixing now.

File: gdal/configure.in (hunk 1 above)
Comment (rouault) [inline gdal/configure.in]: I didn't really understand your answer. This code is meant as detecting if we can enable -Wimplicit-fallthrough safely. For that we need to have both C++11 and Clang 3.7 otherwise the CPL_FALLTROUGH macro will not be defined, and thus warnings would be raised

**Comment by rouault** [comment]
> I test this PR on my ubuntu 16.10. Also clang from 3.3 (this travis-ci report clang version 3.4) also support c++11. There is errors in my configure.in which I'm fixing now.

Anyway, you can merge and then monitor the various builds at https://travis-ci.org/rouault/gdal_coverage/builds/ to add extra corrections if needed

File: gdal/configure.in (hunk 1 above)
Comment (BishopGIS) [inline gdal/configure.in]: There is such error message in code: not C++11 or clang >= 3.7
This makes me misunderstood - as g++ also support c++11. 
Maybe error message should be: not C++11 **and** clang >= 3.7 ?

Comment (rouault) [inline gdal/configure.in]: Ah, you're right the error message is incorrect. It should be not (C++11 and clang >= 3.7) if we want to be pedantic.
The C++11 check is necessary because even if clang >= 3.7 is used, C++11 is not necessarily the default standard used.

**Comment by rouault** [comment]
Regarding the crash in monogdb tests in precise_clang might be due to turning on C++11 while boost has been compiled without C++11 support. See https://mongodb.github.io/mongo-cxx-driver/legacy-v1/installation/#compile-the-driver warnings. Can we have an option to disable C++11 compilation even if the compiler supports it ?
Regarding the failure in  the big_endian test, it was in trunk when you forked it. Rebasing should fix that

**Comment by BishopGIS** [comment]
Thanks Even. This is very helpful. 

What about C++11 optionally use in different drivers - this was done initially by Alexander. But Kurt said to use m4 scripts (ax_cxx_compile_stdcxx and ax_cxx_compile_stdcxx_11). This scripts enable c++11 flags globally, but they rather intelligent to choose supported flags for different compilers, i.e.  -std=gnu++11  or  -std=c++11  or etc.

And now I think what is best solution: 
* fix GDAL to build with c++11 totally, or
* partially build some drivers with c++11

**Comment by rouault** [comment]
@schwehr See above comments from Dmitry in case you have comments

Regarding "fix GDAL to build with c++11 totally", this isn't really a problem of source code compatibility since we have a test target that is already C++11 (and another one C++14). Here the problem is that the precise_clang target uses the system Boost from Precise that isn't likely compatible with C++11 (and the build I did of the mongocxx driver wasn't probably). And as far as I can see, this is the only target that exercices the mongodb driver. So perhaps modify its definition to use an older gcc that isn't C++11 enabled. Or remove the mongodb from that build and consider later enabling it in another build
Actually, looking at my coverage branch ( https://travis-ci.org/rouault/gdal_coverage/jobs/173693043 ) that uses g++ and is close to the lucid_clang, I see it is gcc 4.6, so probably not C++11 OK (but might have partial C++0X, so not sure if it is going to be recognized as C++11 capable)

Regarding partial builds of some drivers with C++11, that might perhaps work (still wondering if that might not bring some ABI issues), but perhaps not worth the complication.

**Comment by BishopGIS** [comment]
The big endian CI failed because https://github.com/OSGeo/gdal/blob/trunk/gdal/frmts/gtiff/geotiff.cpp#L1341
Why this check only on C++11?

**Comment by BishopGIS** [comment]
Merge to svn in r36306
